%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax
\ifdefined\pdfimageresolution
    \pdfimageresolution= \numexpr \dimexpr1in\relax/\sphinxpxdimen\relax
\fi
%% let collapsible pdf bookmarks panel have high depth per default
\PassOptionsToPackage{bookmarksdepth=5}{hyperref}

\PassOptionsToPackage{booktabs}{sphinx}
\PassOptionsToPackage{colorrows}{sphinx}

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{tgtermes}
\usepackage{tgheros}
\renewcommand{\ttdefault}{txtt}



\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=auto}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}

\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\usepackage{sphinxmessages}
\setcounter{tocdepth}{1}



\title{ember\_py}
\date{Nov 19, 2025}
\release{0.1.0}
\author{Nikhila P.\@{} Swarna}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\ifdefined\shorthandoff
  \ifnum\catcode`\=\string=\active\shorthandoff{=}\fi
  \ifnum\catcode`\"=\active\shorthandoff{"}\fi
\fi

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}


\sphinxAtStartPar
Add your content using \sphinxcode{\sphinxupquote{reStructuredText}} syntax. See the
\sphinxhref{https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html}{reStructuredText}
documentation for details.

\sphinxstepscope


\chapter{ember\_py package}
\label{\detokenize{ember_py:ember-py-package}}\label{\detokenize{ember_py::doc}}

\section{Submodules}
\label{\detokenize{ember_py:submodules}}

\section{ember\_py.light\_ember module}
\label{\detokenize{ember_py:module-ember_py.light_ember}}\label{\detokenize{ember_py:ember-py-light-ember-module}}\index{module@\spxentry{module}!ember\_py.light\_ember@\spxentry{ember\_py.light\_ember}}\index{ember\_py.light\_ember@\spxentry{ember\_py.light\_ember}!module@\spxentry{module}}\index{light\_ember() (in module ember\_py.light\_ember)@\spxentry{light\_ember()}\spxextra{in module ember\_py.light\_ember}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.light_ember.light_ember}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.light\_ember.}}\sphinxbfcode{\sphinxupquote{light\_ember}}}
{\sphinxparam{\DUrole{n}{h5ad\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{sampling}\DUrole{o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n}{sample\_id\_col}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{category\_col}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{condition\_col}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{num\_draws}\DUrole{o}{=}\DUrole{default_value}{100}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_draws}\DUrole{o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n}{seed}\DUrole{o}{=}\DUrole{default_value}{42}}\sphinxparamcomma \sphinxparam{\DUrole{n}{partition\_pvals}\DUrole{o}{=}\DUrole{default_value}{True}}\sphinxparamcomma \sphinxparam{\DUrole{n}{block\_pvals}\DUrole{o}{=}\DUrole{default_value}{False}}\sphinxparamcomma \sphinxparam{\DUrole{n}{block\_label}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{n\_pval\_iterations}\DUrole{o}{=}\DUrole{default_value}{1000}}\sphinxparamcomma \sphinxparam{\DUrole{n}{n\_cpus}\DUrole{o}{=}\DUrole{default_value}{1}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Runs the ember entropy metrics and p\sphinxhyphen{}value generation workflow on an AnnData object.

\sphinxAtStartPar
This function loads an AnnData \sphinxtitleref{.h5ad} file, optionally performs balanced sampling
across replicates, computes entropy metrics for the specified partition,
and generates p\sphinxhyphen{}values for Psi and Zeta and optionally Psi\_block for a block of choice.
\begin{description}
\sphinxlineitem{Entropy metrics generated:}\begin{itemize}
\item {} 
\sphinxAtStartPar
Psi : Fraction of infromation explained by partition of choice

\item {} 
\sphinxAtStartPar
Psi\_block : Specificity of infromation to a block

\item {} 
\sphinxAtStartPar
Zeta : Speicifcty to a partition/ distance of Psi\_blocks distribution from uniform

\end{itemize}

\end{description}
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{h5ad\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the \sphinxtitleref{.h5ad} file to process.
Data should be log1p and depth normalized before running ember.
Remove genes with less than 100 reads before running ember.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Column in \sphinxtitleref{.obs} used to partition cells for entropy calculations
(e.g., “celltype”, “Genotype”, “Age”). Required to run process.
If performing calculation on interaction term, first create a column
in \sphinxtitleref{.obs} concatnating the two columns of interested with a semicolon (:).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to directory where results will be saved. Required to run process.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{sampling}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=True}}) \textendash{} Whether to perform balanced sampling across replicates before entropy calculation.
If True, \sphinxtitleref{sample\_id\_col}, \sphinxtitleref{category\_col}, and \sphinxtitleref{condition\_col} must be provided.
Sampling should only be False if fast intermediate results are desired or
if there are no replicates to sample over.
If sampling is set to False but either partition\_pvals or block\_pvals are set to
True then the sampling=False will be overridden as pval generation requires sampling.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{sample\_id\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = None}}) \textendash{} The column in \sphinxtitleref{.obs} with unique identifiers for each sample or replicate
(e.g., ‘sample\_id’, ‘mouse\_id’).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{category\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = None}}) \textendash{} The column in \sphinxtitleref{.obs} defining the primary group to balance across in order
to generate a balanced sample of the experiment. (e.g., ‘disease\_status’, ‘mouse\_strain’).
Refer to readme for further explanation on how to select category and condition columns.
category\_col and condition\_col are interchangable.
If balancing across more than 2 variables, generate interaction terms, create a column
in \sphinxtitleref{.obs} concatnating the two (or more) columns of interested with a semicolon (:).
This way balancing can be done across as many variables as desired.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{condition\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = None}}) \textendash{} The column in \sphinxtitleref{.obs} containing the conditions to balance within
each categoryto generate a balanced sample of the experiment.  (e.g., ‘sex’, ‘treatment’).
Refer to readme for further explanation on how to select category and condition columns.
category\_col and condition\_col are interchangable.
If balancing across more than 2 variables, generate interaction terms, create a column
in \sphinxtitleref{.obs} concatnating the two (or more) columns of interested with a semicolon (:).
This way balancing can be done across as many variables as desired.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{num\_draws}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 100}}) \textendash{} The number of balanced subsets to generate, by default 100.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_draws}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=False}}) \textendash{} Whether to save intermediate draws to save\_dir.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{seed}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{,  }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 42}}) \textendash{} The random seed for reproducible draws, by default 42.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_pvals}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=True}}) \textendash{} Whether to compute permutation\sphinxhyphen{}based p\sphinxhyphen{}values for the \sphinxtitleref{partition\_label}.
P\sphinxhyphen{}values are generated by sampling. If sampling = False and partition\_pvals = True,
the sampling=False will be overwritten.
Calls generate\_pavls, which can be called manually after metric generation as well.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{block\_pvals}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=False}}) \textendash{} Whether to compute permutation\sphinxhyphen{}based p\sphinxhyphen{}values for the \sphinxtitleref{block\_label}.
P\sphinxhyphen{}values are generated by sampling. If sampling = False and block\_pvals = True,
the sampling=False will be overwritten.
Calls generate\_pavls, which can be called manually after metric generation as well.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{block\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,  }}\sphinxstyleliteralemphasis{\sphinxupquote{default = None}}) \textendash{} One value in the \sphinxtitleref{.obs} column for partition\_label to use for block\sphinxhyphen{}based permutation tests.
Required if \sphinxtitleref{block\_pvals=True}.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{n\_pval\_iterations}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=1000}}) \textendash{} Number of permutations to use for p\sphinxhyphen{}value calculation.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{n\_cpus}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=1}}) \textendash{} Number of CPU cores to use for parallel permutation testing.
For this script, performance is I/O\sphinxhyphen{}bound and may not improve beyond 4\sphinxhyphen{}8 cores.’

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}
\subsubsection*{Notes}
\begin{itemize}
\item {} 
\sphinxAtStartPar
Results are saved to \sphinxtitleref{save\_dir} as CSV files.

\item {} 
\sphinxAtStartPar
one csv file with all entropy metrics

\item {} 
\sphinxAtStartPar
one csv file in a new Psi\_block\_df folder with psi block values for all blocks in a partition

\item {} 
\sphinxAtStartPar
Separate file for pvals

\item {} 
\sphinxAtStartPar
Separate files for each partition

\item {} 
\sphinxAtStartPar
Alternate file names depending on sampling on or off.

\end{itemize}

\sphinxAtStartPar
\sphinxstylestrong{What to expect inside ‘entropy\_metrics.csv’}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
gene\_name: All genes in \sphinxtitleref{.var}

\item {} 
\sphinxAtStartPar
Psi\_mean: Psi scores averaged over n draws (between 0 and 1) corresponding to the selected partition for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi\_std: Standard deviation of Psi scores across n draws corresponding to the selected partition for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi\_valid\_counts: Number of valid Psi scores observed across n draws. Only use genes for downstream analysis that have valid counts=num\_draws. If valid counts is not close to num\_draws, increase threshold for filtering genes with low reads beforehand(recommended \textless{}100 reads, increase as needed).

\item {} 
\sphinxAtStartPar
Zeta\_mean: Zeta scores averaged over n draws (between 0 and 1) corresponding to the selected partition for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta\_std: Standard deviation of Zeta scores across n draws corresponding to the selected partition for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta\_valid\_counts: Number of valid Psi scores observed across n draws. Only use genes for downstream analysis that have valid counts=num\_draws. If valid counts is not close to num\_draws, increase threshold for filtering genes with low reads beforehand (recommended \textless{}100 reads, increase as needed).

\end{itemize}

\sphinxAtStartPar
\sphinxstylestrong{What to expect inside ‘Psi\_block\_df/’}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
mean\_Psi\_block\_df.csv : A dataframe of mean Psi\_block scores (between 0 and 1) corresponding to the selected partition for each gene in \sphinxtitleref{.var}. Scores are caluclated for all blocks, each column of the dataframe corresponds to one block.

\item {} 
\sphinxAtStartPar
std\_Psi\_block\_df.csv : A dataframe of standard deviations for Psi\_block scores corresponding to the selected partition for each gene in \sphinxtitleref{.var}.Scores are caluclated for all blocks, each column of the dataframe corresponds to one block.

\end{itemize}

\sphinxAtStartPar
\sphinxstylestrong{What to expect inside ‘pvals\_entropy\_metrics.csv’}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
gene\_name: All genes in \sphinxtitleref{.var}

\item {} 
\sphinxAtStartPar
Psi: Psi scores averaged over n draws (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed Psi scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta: Zeta scores averaged over n draws (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed Zeta scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for Psi scores.

\item {} 
\sphinxAtStartPar
Zeta q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for Zeta scores.Correction perfromed to include all p\sphinxhyphen{}values generated in a single file (Psi and Zeta).

\end{itemize}

\sphinxAtStartPar
If block\_pvals = True and a single block\_label is given:
\begin{itemize}
\item {} 
\sphinxAtStartPar
psi\_block: psi\_block scores (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
psi\_block p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed psi\_block scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
psi\_block q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for psi\_block scores. Correction perfromed to include all p\sphinxhyphen{}values generated in a single file (Psi, psi\_block and Zeta).

\end{itemize}

\end{fulllineitems}



\section{ember\_py.generate\_pvals module}
\label{\detokenize{ember_py:module-ember_py.generate_pvals}}\label{\detokenize{ember_py:ember-py-generate-pvals-module}}\index{module@\spxentry{module}!ember\_py.generate\_pvals@\spxentry{ember\_py.generate\_pvals}}\index{ember\_py.generate\_pvals@\spxentry{ember\_py.generate\_pvals}!module@\spxentry{module}}\index{generate\_pvals() (in module ember\_py.generate\_pvals)@\spxentry{generate\_pvals()}\spxextra{in module ember\_py.generate\_pvals}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.generate_pvals.generate_pvals}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.generate\_pvals.}}\sphinxbfcode{\sphinxupquote{generate\_pvals}}}
{\sphinxparam{\DUrole{n}{h5ad\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{entropy\_metrics\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{sample\_id\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{category\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{condition\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{block\_label}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{seed}\DUrole{o}{=}\DUrole{default_value}{42}}\sphinxparamcomma \sphinxparam{\DUrole{n}{n\_iterations}\DUrole{o}{=}\DUrole{default_value}{1000}}\sphinxparamcomma \sphinxparam{\DUrole{n}{n\_cpus}\DUrole{o}{=}\DUrole{default_value}{1}}\sphinxparamcomma \sphinxparam{\DUrole{n}{Psi\_real}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{Psi\_block\_df\_real}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{Zeta\_real}\DUrole{o}{=}\DUrole{default_value}{None}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Calculate empirical p\sphinxhyphen{}values for entropy metrics from permutation test results.
This function can be called manually or accessed through light\_ember with
partition\_pvals = True or block\_pvals = True.

\sphinxAtStartPar
Manual access useful if wanting to generate p\sphinxhyphen{}values for multiple blocks and partitions of
interest after initial investigation using entropy metrics.

\sphinxAtStartPar
Integrated access with light\_ember is easier if investigating only a partition or
a block in a partition.
\begin{description}
\sphinxlineitem{Entropy metrics generated:}\begin{itemize}
\item {} 
\sphinxAtStartPar
Psi : Fraction of infromation explained by partition of choice

\item {} 
\sphinxAtStartPar
Psi\_block : Specificity of infromation to a block

\item {} 
\sphinxAtStartPar
Zeta : Speicifcty to a partition/ distance of Psi\_blocks distribution from uniform

\end{itemize}

\end{description}
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{h5ad\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the \sphinxtitleref{.h5ad} file to process.
Data should be log1p and depth normalized before running ember.
Remove genes with less than 100 reads before running ember.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Column in \sphinxtitleref{.obs} used to partition cells for entropy calculations
(e.g., “celltype”, “Genotype”, “Age”). Required to run process.
If performing calculation on interaction term, first create a column
in \sphinxtitleref{.obs} concatnating the two columns of interested with a semicolon (:).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{entropy\_metrics\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to csv with entropy metrics to use for generating pvals.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to directory where results will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{sample\_id\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column in \sphinxtitleref{.obs} with unique identifiers for each sample or replicate
(e.g., ‘sample\_id’, ‘mouse\_id’).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{category\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column in \sphinxtitleref{.obs} defining the primary group to balance across in order
to generate a balanced sample of the experiment. (e.g., ‘disease\_status’, ‘mouse\_strain’).
Refer to readme for further explanation on how to select category and condition columns.
category\_col and condition\_col are interchangable.
If balancing across more than 2 variables, generate interaction terms, create a column
in \sphinxtitleref{.obs} concatnating the two (or more) columns of interested with a semicolon (:).
This way balancing can be done across as many variables as desired.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{condition\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column in \sphinxtitleref{.obs} containing the conditions to balance within
each category to generate a balanced sample of the experiment.  (e.g., ‘sex’, ‘treatment’).
Refer to readme for further explanation on how to select category and condition columns.
category\_col and condition\_col are interchangable.
If balancing across more than 2 variables, generate interaction terms, create a column
in \sphinxtitleref{.obs} concatnating the two (or more) columns of interested with a semicolon (:).
This way balancing can be done across as many variables as desired.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{block\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None}}) \textendash{} Block in partition to calucate p\sphinxhyphen{}values for.
Default set to None, program will continue generating p\sphinxhyphen{}values for only Psi and Zeta.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{seed}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=42}}) \textendash{} The random seed for reproducible draws, by default 42.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{n\_iterations}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 1000}}) \textendash{} Number of iterations to calulate p\sphinxhyphen{}vales. Default set to 1000.
Note that doing fewer than 1000 iterations is a good choice to get first pass p\sphinxhyphen{}values
but for reliable p\sphinxhyphen{}values 1000 iterations is recommended.
Larger than 1000 will give you more relibale p\sphinxhyphen{}values but will increase runtime significantly.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{n\_cpus}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=1}}) \textendash{} Number of cpus to use to perfrom p\sphinxhyphen{}value calculation.
Default set to 1 assuming no parallel compute power on local machine.
User can input \sphinxhyphen{}1 to use all available cpus but one.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{Psi\_real}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.Series}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None}}) \textendash{} Observed Psi values for each gene.
Used by light\_ember, not necessary for user use.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{Psi\_block\_df\_real}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.Dataframe}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = None}}) \textendash{} Observed Psi\_block values for all blocks in chosen partition.
Used by light\_ember, not necessary for user use.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{Zeta\_real}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.Series}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None}}) \textendash{} Observed Zeta values for each gene.
Used by light\_ember, not necessary for user use.

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}
\subsubsection*{Notes}

\sphinxAtStartPar
\sphinxstylestrong{What to expect inside ‘pvals\_entropy\_metrics.csv’}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
gene\_name: All genes in \sphinxtitleref{.var}

\item {} 
\sphinxAtStartPar
Psi: Psi scores averaged over n draws (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed Psi scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta: Zeta scores averaged over n draws (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Zeta p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed Zeta scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
Psi q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for Psi scores.

\item {} 
\sphinxAtStartPar
Zeta q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for Zeta scores.Correction perfromed to include all p\sphinxhyphen{}values generated in a single file (Psi and Zeta).

\end{itemize}

\sphinxAtStartPar
if block\_pvals = True and a single block\_label is given:
\begin{itemize}
\item {} 
\sphinxAtStartPar
psi\_block: psi\_block scores (between 0 and 1) generated by light\_ember for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
psi\_block p\sphinxhyphen{}value: Permutation based empirical p\sphinxhyphen{}values for observed psi\_block scores for each gene in \sphinxtitleref{.var}.

\item {} 
\sphinxAtStartPar
psi\_block q\sphinxhyphen{}value: Multiple testing corrected q\sphinxhyphen{}values for psi\_block scores. Correction perfromed to include all p\sphinxhyphen{}values generated in a single file (Psi, psi\_block and Zeta).

\end{itemize}

\end{fulllineitems}



\section{ember\_py.plots module}
\label{\detokenize{ember_py:module-ember_py.plots}}\label{\detokenize{ember_py:ember-py-plots-module}}\index{module@\spxentry{module}!ember\_py.plots@\spxentry{ember\_py.plots}}\index{ember\_py.plots@\spxentry{ember\_py.plots}!module@\spxentry{module}}\index{plot\_block\_specificity() (in module ember\_py.plots)@\spxentry{plot\_block\_specificity()}\spxextra{in module ember\_py.plots}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.plots.plot_block_specificity}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.plots.}}\sphinxbfcode{\sphinxupquote{plot\_block\_specificity}}}
{\sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{block\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{pvals\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{highlight\_genes}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{q\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.05}}\sphinxparamcomma \sphinxparam{\DUrole{n}{fontsize}\DUrole{o}{=}\DUrole{default_value}{18}}\sphinxparamcomma \sphinxparam{\DUrole{n}{custom\_palette}\DUrole{o}{=}\DUrole{default_value}{None}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Generate a psi\_block vs. Psi scatter plot to visualize block\sphinxhyphen{}specific genes.

\sphinxAtStartPar
This function reads p\sphinxhyphen{}value data, colors genes based on their statistical
significance for Psi and psi\_block scores, and highlights the top genes
significant in both metrics. Only interpret genes that are significant by both Psi
and psi\_block since those are genes that have reliable scores after permutation
testing.
Allows for custom highlighting of a user\sphinxhyphen{}provided gene list.
Fontsize and color pallette can be customized.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} The label for the partition, used in the plot title.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{block\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} The label for the block variable (e.g., a cell type or condition).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{pvals\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} Path to the input CSV file containing p\sphinxhyphen{}values and scores.
The CSV must have gene names as its index column.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} Path where the output plot image will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{highlight\_genes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None.}}) \textendash{} A list of gene names to highlight and annotate on the plot, by default None.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{q\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.05}}) \textendash{} Threshold for q\sphinxhyphen{}values. Genes are retained if both
‘Psi q\sphinxhyphen{}value’ \textless{}= q\_thresh and ‘psi\_block q\sphinxhyphen{}value’ \textless{}= q\_thresh.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{fontsize}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 18.}}) \textendash{} The base font size for plot labels and text, by default 18.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{custom\_palette}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None.}}) \textendash{} A list of 6 hex color codes to customize the plot’s color scheme.
If None, a default palette is used.
Provide list of colors int his order:
{[}‘significant by psi’,
‘significant by psi\_block’,
‘highlight genes’,
‘significant by both’,
‘cirlce markers’,
‘circle housekeeping genes’,
‘significant by neither’{]}

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_partition\_specificity() (in module ember\_py.plots)@\spxentry{plot\_partition\_specificity()}\spxextra{in module ember\_py.plots}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.plots.plot_partition_specificity}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.plots.}}\sphinxbfcode{\sphinxupquote{plot\_partition\_specificity}}}
{\sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{pvals\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{highlight\_genes}\DUrole{o}{=}\DUrole{default_value}{None}}\sphinxparamcomma \sphinxparam{\DUrole{n}{q\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.05}}\sphinxparamcomma \sphinxparam{\DUrole{n}{fontsize}\DUrole{o}{=}\DUrole{default_value}{18}}\sphinxparamcomma \sphinxparam{\DUrole{n}{custom\_palette}\DUrole{o}{=}\DUrole{default_value}{None}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Generate a Zeta vs. Psi scatter plot to visualize partition\sphinxhyphen{}specific genes.

\sphinxAtStartPar
This function reads p\sphinxhyphen{}value data, colors genes based on their statistical
significance for Psi and Zeta scores, and highlights top “marker” and
“housekeeping” genes. Only interpret genes that are significant by both Psi
and Zeta since those are genes that have reliable scores after permutation
testing.
Allows for custom highlighting of a user\sphinxhyphen{}provided gene list.
Fontsize and color pallette can be customized.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} The label for the partition being plotted, used in the plot title.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{pvals\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} Path to the input CSV file containing p\sphinxhyphen{}values and scores (Psi, Zeta, q\sphinxhyphen{}values).
The CSV must have gene names as its index column.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required.}}) \textendash{} Path where the output plot image will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{highlight\_genes}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None.}}) \textendash{} A list of gene names to highlight and annotate on the plot, by default None.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{q\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.05}}) \textendash{} Threshold for q\sphinxhyphen{}values. Genes are retained if both
‘Psi q\sphinxhyphen{}value’ \textless{}= q\_thresh and ‘Zeta q\sphinxhyphen{}value’ \textless{}= q\_thresh.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{fontsize}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=18.}}) \textendash{} The base font size for plot labels and text, by default 18.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{custom\_palette}} (\sphinxstyleliteralemphasis{\sphinxupquote{list}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=None.}}) \textendash{} A list of 7 hex color codes to customize the plot’s color scheme.
If None, a default palette is used.
Please provide list in this order
{[}‘significant by psi’,
‘significant by zeta’,
‘highlight genes’,
‘significant by both’,
‘cirlce markers’,
‘circle housekeeping genes’,
‘significant by neither’{]}

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_psi\_blocks() (in module ember\_py.plots)@\spxentry{plot\_psi\_blocks()}\spxextra{in module ember\_py.plots}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.plots.plot_psi_blocks}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.plots.}}\sphinxbfcode{\sphinxupquote{plot\_psi\_blocks}}}
{\sphinxparam{\DUrole{n}{gene\_name}}\sphinxparamcomma \sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{psi\_block\_df\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{fontsize}\DUrole{o}{=}\DUrole{default_value}{18}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Generates and saves a bar plot of mean psi block values with error bars.

\sphinxAtStartPar
This function reads two CSV files from a specified directory: one for mean
psi block values and one for standard deviations. It plots the mean values
for a specific gene as a bar plot with corresponding standard deviation
error bars.
Fontsize can be customized.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The name of the gene (row) to select and plot from the CSV files.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The partition label used to find the correct files (e.g., ‘Genotype’).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{psi\_block\_df\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the directory containing the mean and std CSV files. Files must
be named ‘mean\_Psi\_block\_df\_\{partition\_label\}.csv’ and
‘std\_Psi\_block\_df\_\{partition\_label\}.csv’.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to directory to save the output plot image.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{fontsize}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default=18.}}) \textendash{} The base font size for plot labels and text, by default 18.

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_sample\_counts() (in module ember\_py.plots)@\spxentry{plot\_sample\_counts()}\spxextra{in module ember\_py.plots}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.plots.plot_sample_counts}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.plots.}}\sphinxbfcode{\sphinxupquote{plot\_sample\_counts}}}
{\sphinxparam{\DUrole{n}{h5ad\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{sample\_id\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{category\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{condition\_col}}\sphinxparamcomma \sphinxparam{\DUrole{n}{fontsize}\DUrole{o}{=}\DUrole{default_value}{18}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Generate a bar plot showing the number of unique individuals
per category and condition.

\sphinxAtStartPar
This function reads an AnnData object from an .h5ad file in backed mode,
calculates the number of unique individuals for each combination of a given
category and condition, and visualizes these counts as a grouped bar plot.
Fontsize can be customized.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{h5ad\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the input AnnData (.h5ad) file.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to directory to save the output plot image.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{sample\_id\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column name in adata.obs that contains unique sample IDs.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{category\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column name to use for the primary categories on the x\sphinxhyphen{}axis.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{condition\_col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} The column name to use for grouping the bars (hue).

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{fontsize}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 18.}}) \textendash{} The base font size for plot labels and text, by default 18.

\end{itemize}

\sphinxlineitem{Return type}
\sphinxAtStartPar
None

\end{description}\end{quote}

\end{fulllineitems}



\section{ember\_py.top\_genes module}
\label{\detokenize{ember_py:module-ember_py.top_genes}}\label{\detokenize{ember_py:ember-py-top-genes-module}}\index{module@\spxentry{module}!ember\_py.top\_genes@\spxentry{ember\_py.top\_genes}}\index{ember\_py.top\_genes@\spxentry{ember\_py.top\_genes}!module@\spxentry{module}}\index{highly\_specific\_to\_block() (in module ember\_py.top\_genes)@\spxentry{highly\_specific\_to\_block()}\spxextra{in module ember\_py.top\_genes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.top_genes.highly_specific_to_block}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.top\_genes.}}\sphinxbfcode{\sphinxupquote{highly\_specific\_to\_block}}}
{\sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{block\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{pvals\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{psi\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{psi\_block\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{q\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.05}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Identifies significant and specific genes from a ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, psi\_block, and q\sphinxhyphen{}values.
(Potential marker genes)

\sphinxAtStartPar
This function reads a CSV file containing Psi and psi\_block metrics (and their
corresponding q\sphinxhyphen{}values), filters genes that meet given significance and
specificity thresholds, and saves the resulting subset to a new CSV file.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{pvals\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the input CSV file (e.g., ‘pvals\_entropy\_metrics\_Age\_E16.5.csv’).
The CSV must include the following columns:
‘Psi q\sphinxhyphen{}value’, ‘psi\_block q\sphinxhyphen{}value’, ‘Psi’, and ‘psi\_block’.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Directory where the filtered results CSV will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Name of partition used to generate entropy metrics, used to label saved csv.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{block\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Name of block in partition used to generate entropy metrics, used to label saved csv.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{psi\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for Psi values. Only genes with Psi \textgreater{} psi\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{psi\_block\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for psi\_block values. Only genes with psi\_block \textgreater{} psi\_block\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{q\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.05}}) \textendash{} Threshold for q\sphinxhyphen{}values. Genes are retained if both
‘Psi q\sphinxhyphen{}value’ \textless{}= q\_thresh and ‘psi\_block q\sphinxhyphen{}value’ \textless{}= q\_thresh.

\end{itemize}

\sphinxlineitem{Returns}
\sphinxAtStartPar
DataFrame containing the significant and specific genes that meet
all threshold criteria. Also saved as
“highly\_specific\_genes\_by\_\{partition\_label\}\_\{block\_label\}.csv”
in the specified save directory.

\sphinxlineitem{Return type}
\sphinxAtStartPar
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{highly\_specific\_to\_partition() (in module ember\_py.top\_genes)@\spxentry{highly\_specific\_to\_partition()}\spxextra{in module ember\_py.top\_genes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.top_genes.highly_specific_to_partition}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.top\_genes.}}\sphinxbfcode{\sphinxupquote{highly\_specific\_to\_partition}}}
{\sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{pvals\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{psi\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{zeta\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{q\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.05}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Identifies significant and specific genes from a ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, Zeta, and q\sphinxhyphen{}values.

\sphinxAtStartPar
This function reads a CSV file containing Psi and Zeta metrics (and their
corresponding q\sphinxhyphen{}values), filters genes that meet given significance and
specificity thresholds, and saves the resulting subset to a new CSV file.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{pvals\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the input CSV file (e.g., ‘pvals\_entropy\_metrics\_Age\_E16.5.csv’).
The CSV must include the following columns:
‘Psi q\sphinxhyphen{}value’, ‘Zeta q\sphinxhyphen{}value’, ‘Psi’, and ‘Zeta’.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Directory where the filtered results CSV will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Name of partition used to generate entropy metrics, used to label saved csv.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{psi\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for Psi values. Only genes with Psi \textgreater{} psi\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{zeta\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for Zeta values. Only genes with Zeta \textgreater{} zeta\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{q\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.05}}) \textendash{} Threshold for q\sphinxhyphen{}values. Genes are retained if both
‘Psi q\sphinxhyphen{}value’ \textless{}= q\_thresh and ‘Zeta q\sphinxhyphen{}value’ \textless{}= q\_thresh.

\end{itemize}

\sphinxlineitem{Returns}
\sphinxAtStartPar
DataFrame containing the significant and specific genes that meet
all threshold criteria. Also saved as “highly\_specific\_genes\_to\_\{partition\_label\}.csv”
in the specified save directory.

\sphinxlineitem{Return type}
\sphinxAtStartPar
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{non\_specific\_to\_partition() (in module ember\_py.top\_genes)@\spxentry{non\_specific\_to\_partition()}\spxextra{in module ember\_py.top\_genes}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{ember_py:ember_py.top_genes.non_specific_to_partition}}
\pysigstartsignatures
\pysiglinewithargsret
{\sphinxcode{\sphinxupquote{ember\_py.top\_genes.}}\sphinxbfcode{\sphinxupquote{non\_specific\_to\_partition}}}
{\sphinxparam{\DUrole{n}{partition\_label}}\sphinxparamcomma \sphinxparam{\DUrole{n}{pvals\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{save\_dir}}\sphinxparamcomma \sphinxparam{\DUrole{n}{psi\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{zeta\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.5}}\sphinxparamcomma \sphinxparam{\DUrole{n}{q\_thresh}\DUrole{o}{=}\DUrole{default_value}{0.05}}}
{}
\pysigstopsignatures
\sphinxAtStartPar
Identifies significant and non\sphinxhyphen{}specific genes from a ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, Zeta, and q\sphinxhyphen{}values.
(Potential housekeeping genes)

\sphinxAtStartPar
This function reads a CSV file containing Psi and Zeta metrics (and their
corresponding q\sphinxhyphen{}values), filters genes that meet given significance and
specificity thresholds, and saves the resulting subset to a new CSV file.
\begin{quote}\begin{description}
\sphinxlineitem{Parameters}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{pvals\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Path to the input CSV file (e.g., ‘pvals\_entropy\_metrics\_Age\_E16.5.csv’).
The CSV must include the following columns:
‘Psi q\sphinxhyphen{}value’, ‘Zeta q\sphinxhyphen{}value’, ‘Psi’, and ‘Zeta’.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{save\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Directory where the filtered results CSV will be saved.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{partition\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{Required}}) \textendash{} Name of partition used to generate entropy metrics, used to label saved csv.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{psi\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for Psi values. Only genes with Psi \textgreater{} psi\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{zeta\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.5}}) \textendash{} Threshold for Zeta values. Only genes with Zeta \textless{} zeta\_thresh are kept.

\item {} 
\sphinxAtStartPar
\sphinxstyleliteralstrong{\sphinxupquote{q\_thresh}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{default = 0.05}}) \textendash{} Threshold for q\sphinxhyphen{}values. Genes are retained if both
‘Psi q\sphinxhyphen{}value’ \textless{}= q\_thresh and ‘Zeta q\sphinxhyphen{}value’ \textless{}= q\_thresh.

\end{itemize}

\sphinxlineitem{Returns}
\sphinxAtStartPar
DataFrame containing the significant and specific genes that meet
all threshold criteria. Also saved as “non\_specific\_genes\_to\_\{partition\_label\}.csv”
in the specified save directory.

\sphinxlineitem{Return type}
\sphinxAtStartPar
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}



\section{Module contents}
\label{\detokenize{ember_py:module-ember_py}}\label{\detokenize{ember_py:module-contents}}\index{module@\spxentry{module}!ember\_py@\spxentry{ember\_py}}\index{ember\_py@\spxentry{ember\_py}!module@\spxentry{module}}
\sphinxstepscope


\chapter{Command Line Interface}
\label{\detokenize{cli:command-line-interface}}\label{\detokenize{cli::doc}}
\sphinxAtStartPar
The \sphinxcode{\sphinxupquote{ember}} toolkit provides a command\sphinxhyphen{}line interface with multiple
subcommands for computing entropy metrics, generating p\sphinxhyphen{}values, plotting
summaries, and extracting highly specific or non\sphinxhyphen{}specific genes.

\sphinxAtStartPar
This page documents the full CLI using the same argparse parser that the
tool uses internally.

\sphinxAtStartPar

\sphinxAtStartPar
A command\sphinxhyphen{}line toolkit for ember: Entropy Metrics for Biological ExploRation.


\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{ember} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]}
             \PYG{p}{\PYGZob{}}\PYG{n}{light\PYGZus{}ember}\PYG{p}{,}\PYG{n}{generate\PYGZus{}pvals}\PYG{p}{,}\PYG{n}{plot\PYGZus{}partition\PYGZus{}specificity}\PYG{p}{,}\PYG{n}{plot\PYGZus{}block\PYGZus{}specificity}\PYG{p}{,}\PYG{n}{plot\PYGZus{}sample\PYGZus{}counts}\PYG{p}{,}\PYG{n}{plot\PYGZus{}psi\PYGZus{}blocks}\PYG{p}{,}\PYG{n}{highly\PYGZus{}specific\PYGZus{}to\PYGZus{}partition}\PYG{p}{,}\PYG{n}{highly\PYGZus{}specific\PYGZus{}to\PYGZus{}block}\PYG{p}{,}\PYG{n}{non\PYGZus{}specific\PYGZus{}to\PYGZus{}partition}\PYG{p}{\PYGZcb{}}
             \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}
\end{sphinxVerbatim}


\section{Positional Arguments}
\label{\detokenize{cli:ember_py.cli-create_parser-positional-arguments}}\begin{optionlist}{3cm}
\item [command]  
\sphinxAtStartPar
Possible choices: light\_ember, generate\_pvals, plot\_partition\_specificity, plot\_block\_specificity, plot\_sample\_counts, plot\_psi\_blocks, highly\_specific\_to\_partition, highly\_specific\_to\_block, non\_specific\_to\_partition

\sphinxAtStartPar
Available sub\sphinxhyphen{}commands
\end{optionlist}


\section{Sub\sphinxhyphen{}commands}
\label{\detokenize{cli:Sub-commands}}

\subsection{light\_ember}
\label{\detokenize{cli:light_ember}}\begin{quote}

\sphinxAtStartPar
Runs the ember entropy metrics and p\sphinxhyphen{}value generation workflow on an AnnData object.

\sphinxAtStartPar
This function loads an AnnData \sphinxtitleref{.h5ad} file, optionally performs balanced sampling
across replicates, computes entropy metrics for the specified partition,
and generates p\sphinxhyphen{}values for Psi and Zeta and optionally Psi\_block for a block of choice.
\begin{description}
\sphinxlineitem{Entropy metrics generated:}\begin{itemize}
\item {} 
\sphinxAtStartPar
Psi : Fraction of information explained by partition of choice

\item {} 
\sphinxAtStartPar
Psi\_block : Specificity of information to a block

\item {} 
\sphinxAtStartPar
Zeta : Specificity to a partition / distance of Psi\_blocks distribution from uniform

\end{itemize}

\end{description}

\sphinxAtStartPar
Notes:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Results are saved to \sphinxtitleref{save\_dir} as CSV files.

\item {} 
\sphinxAtStartPar
One CSV file with all entropy metrics.

\item {} 
\sphinxAtStartPar
One CSV file in a new Psi\_block\_df folder with Psi\_block values for all blocks in a partition.

\item {} 
\sphinxAtStartPar
Separate file for p\sphinxhyphen{}values.

\item {} 
\sphinxAtStartPar
Separate files for each partition.

\item {} 
\sphinxAtStartPar
Alternate file names depending on sampling on or off.

\end{itemize}
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{light\PYGZus{}ember} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{no\PYGZus{}sampling}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{sample\PYGZus{}id\PYGZus{}col} \PYG{n}{SAMPLE\PYGZus{}ID\PYGZus{}COL}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{category\PYGZus{}col} \PYG{n}{CATEGORY\PYGZus{}COL}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{condition\PYGZus{}col} \PYG{n}{CONDITION\PYGZus{}COL}\PYG{p}{]}
                  \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{num\PYGZus{}draws} \PYG{n}{NUM\PYGZus{}DRAWS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{save\PYGZus{}draws}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{seed} \PYG{n}{SEED}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{no\PYGZus{}partition\PYGZus{}pvals}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{block\PYGZus{}pvals}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{block\PYGZus{}label} \PYG{n}{BLOCK\PYGZus{}LABEL}\PYG{p}{]}
                  \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{n\PYGZus{}pval\PYGZus{}iterations} \PYG{n}{N\PYGZus{}PVAL\PYGZus{}ITERATIONS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{n\PYGZus{}cpus} \PYG{n}{N\PYGZus{}CPUS}\PYG{p}{]}
                  \PYG{n}{h5ad\PYGZus{}dir} \PYG{n}{partition\PYGZus{}label} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments}}\begin{optionlist}{3cm}
\item [h5ad\_dir]  
\sphinxAtStartPar
Path to the \sphinxtitleref{.h5ad} file to process. Data should be log1p and depth normalized before running ember. Remove genes with \textless{}100 reads before running ember.
\item [partition\_label]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} used to partition cells for entropy calculations (e.g., ‘celltype’, ‘Genotype’, ‘Age’). For interaction terms, create a new column concatenating multiple \sphinxtitleref{.obs} columns with a semicolon (:).
\item [save\_dir]  
\sphinxAtStartPar
Path to directory where results will be saved.
\end{optionlist}


\subsubsection{Sampling Parameters}
\label{\detokenize{cli:sampling-parameters}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}no\_sampling]  
\sphinxAtStartPar
Disable balanced sampling. Default: True. Note: If partition\_pvals or block\_pvals are enabled, sampling will be re\sphinxhyphen{}enabled.
\item [\sphinxhyphen{}\sphinxhyphen{}sample\_id\_col]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} with unique identifiers for each sample or replicate (e.g., ‘sample\_id’, ‘mouse\_id’).
\item [\sphinxhyphen{}\sphinxhyphen{}category\_col]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} defining the primary group to balance across (e.g., ‘disease\_status’, ‘mouse\_strain’). Interchangeable with condition\_col. For \textgreater{}2 variables, create interaction terms by concatenating columns with \sphinxtitleref{:}.
\item [\sphinxhyphen{}\sphinxhyphen{}condition\_col]  
\sphinxAtStartPar
Secondary column in \sphinxtitleref{.obs} to balance sampling across (e.g., ‘sex’, ‘treatment’). Interchangeable with category\_col. Supports interaction terms.
\item [\sphinxhyphen{}\sphinxhyphen{}num\_draws]  
\sphinxAtStartPar
Number of balanced subsets to generate (default: 100).
\item [\sphinxhyphen{}\sphinxhyphen{}save\_draws]  
\sphinxAtStartPar
Save intermediate sampled draws to save\_dir (default: False).
\item [\sphinxhyphen{}\sphinxhyphen{}seed]  
\sphinxAtStartPar
Random seed for reproducible draws (default: 42).
\end{optionlist}


\subsubsection{P\sphinxhyphen{}value Parameters}
\label{\detokenize{cli:p-value-parameters}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}no\_partition\_pvals]  
\sphinxAtStartPar
Disable permutation p\sphinxhyphen{}value calculation for the main partition. Default: True.
\item [\sphinxhyphen{}\sphinxhyphen{}block\_pvals]  
\sphinxAtStartPar
Enable permutation p\sphinxhyphen{}value calculation for a specific block. Default: False.
\item [\sphinxhyphen{}\sphinxhyphen{}block\_label]  
\sphinxAtStartPar
Specific value in ‘partition\_label’ for block p\sphinxhyphen{}values. Required if \textendash{}block\_pvals is set.
\item [\sphinxhyphen{}\sphinxhyphen{}n\_pval\_iterations]  
\sphinxAtStartPar
Number of permutations for p\sphinxhyphen{}value calculation (default: 1000).
\end{optionlist}


\subsubsection{Performance Parameters}
\label{\detokenize{cli:performance-parameters}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}n\_cpus]  
\sphinxAtStartPar
Number of CPU cores to use for parallel processing (default: 1). Performance is I/O\sphinxhyphen{}bound and may not improve beyond 4\textendash{}8 cores.
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember light\_ember \textasciitilde{}/ember\_test/test\_adata\_cwc22.h5ad Genotype \textasciitilde{}/ember\_test/ \textendash{}sample\_id\_col Mouse\_ID \textendash{}category\_col Genotype \textendash{}condition\_col Sex \textendash{}num\_draws 50 \textendash{}no\_partition\_pvals \textendash{}n\_cpus 4

\end{description}
\end{quote}


\subsection{generate\_pvals}
\label{\detokenize{cli:generate_pvals}}\begin{quote}

\sphinxAtStartPar
Calculate empirical p\sphinxhyphen{}values for entropy metrics from permutation test results.
\begin{description}
\sphinxlineitem{Entropy metrics generated:}\begin{itemize}
\item {} 
\sphinxAtStartPar
Psi : Fraction of information explained by partition of choice

\item {} 
\sphinxAtStartPar
Psi\_block : Specificity of information to a block

\item {} 
\sphinxAtStartPar
Zeta : Specificity to a partition / distance of Psi\_blocks distribution from uniform

\end{itemize}

\end{description}
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{generate\PYGZus{}pvals} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{block\PYGZus{}label} \PYG{n}{BLOCK\PYGZus{}LABEL}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{seed} \PYG{n}{SEED}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{n\PYGZus{}iterations} \PYG{n}{N\PYGZus{}ITERATIONS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{n\PYGZus{}cpus} \PYG{n}{N\PYGZus{}CPUS}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{Psi\PYGZus{}real} \PYG{n}{PSI\PYGZus{}REAL}\PYG{p}{]}
                     \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{Psi\PYGZus{}block\PYGZus{}df\PYGZus{}real} \PYG{n}{PSI\PYGZus{}BLOCK\PYGZus{}DF\PYGZus{}REAL}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{Zeta\PYGZus{}real} \PYG{n}{ZETA\PYGZus{}REAL}\PYG{p}{]}
                     \PYG{n}{h5ad\PYGZus{}dir} \PYG{n}{partition\PYGZus{}label} \PYG{n}{entropy\PYGZus{}metrics\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir} \PYG{n}{sample\PYGZus{}id\PYGZus{}col} \PYG{n}{category\PYGZus{}col} \PYG{n}{condition\PYGZus{}col}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat1}}\begin{optionlist}{3cm}
\item [h5ad\_dir]  
\sphinxAtStartPar
Path to the \sphinxtitleref{.h5ad} file to process. Data should be log1p and depth normalized before running ember. Remove genes with \textless{}100 reads before running ember.
\item [partition\_label]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} used to partition cells for entropy calculations (e.g., ‘celltype’, ‘Genotype’, ‘Age’). For interaction terms, create a new column concatenating multiple \sphinxtitleref{.obs} columns with a semicolon (:).
\item [entropy\_metrics\_dir]  
\sphinxAtStartPar
Path to CSV with entropy metrics to use for generating p\sphinxhyphen{}values.
\item [save\_dir]  
\sphinxAtStartPar
Path to directory where results will be saved.
\item [sample\_id\_col]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} with unique identifiers for each sample or replicate (e.g., ‘sample\_id’, ‘mouse\_id’).
\item [category\_col]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} defining the primary group to balance across (e.g., ‘disease\_status’, ‘mouse\_strain’). Interchangeable with condition\_col. For \textgreater{}2 variables, create interaction terms by concatenating columns with \sphinxtitleref{:}.
\item [condition\_col]  
\sphinxAtStartPar
Column in \sphinxtitleref{.obs} containing the conditions to balance within each category (e.g., ‘sex’, ‘treatment’). Interchangeable with category\_col. Supports interaction terms.
\end{optionlist}


\subsubsection{Named Arguments}
\label{\detokenize{cli:named-arguments}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}block\_label]  
\sphinxAtStartPar
Block in partition to calculate p\sphinxhyphen{}values for. Default: None (Psi and Zeta only).
\end{optionlist}


\subsubsection{Performance Parameters}
\label{\detokenize{cli:performance-parameters_repeat1}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}seed]  
\sphinxAtStartPar
Random seed for reproducible draws (default: 42).
\item [\sphinxhyphen{}\sphinxhyphen{}n\_iterations]  
\sphinxAtStartPar
Number of iterations to calculate p\sphinxhyphen{}values (default: 1000). Use fewer for quick runs, more for reliable results.
\item [\sphinxhyphen{}\sphinxhyphen{}n\_cpus]  
\sphinxAtStartPar
Number of CPUs to use for p\sphinxhyphen{}value calculation (default: 1). Set to \sphinxhyphen{}1 to use all available cores but one.
\end{optionlist}


\subsubsection{Internal Arguments (used by light\_ember)}
\label{\detokenize{cli:internal-arguments-(used-by-light_ember)}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}Psi\_real]  
\sphinxAtStartPar
Observed Psi values for each gene (pd.Series). Not required for user runs.
\item [\sphinxhyphen{}\sphinxhyphen{}Psi\_block\_df\_real]  
\sphinxAtStartPar
Observed Psi\_block values for all blocks in chosen partition (pd.DataFrame). Not required for user runs.
\item [\sphinxhyphen{}\sphinxhyphen{}Zeta\_real]  
\sphinxAtStartPar
Observed Zeta values for each gene (pd.Series). Not required for user runs.
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember generate\_pvals test\_adata\_cwc22.h5ad Genotype \textasciitilde{}/ember\_test/ \textasciitilde{}/ember\_test/output Mouse\_ID Genotype Sex \textendash{}block\_label WSBJ \textendash{}n\_cpus 4

\end{description}
\end{quote}


\subsection{plot\_partition\_specificity}
\label{\detokenize{cli:plot_partition_specificity}}\begin{quote}

\sphinxAtStartPar
Generate a Zeta vs. Psi scatter plot to visualize partition\sphinxhyphen{}specific genes.

\sphinxAtStartPar
This function reads p\sphinxhyphen{}value data, colors genes based on their statistical
significance for Psi and Zeta scores, and highlights top “marker” and
“housekeeping” genes. Allows for custom highlighting of a user\sphinxhyphen{}provided
gene list. Font size and color palette can be customized.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{plot\PYGZus{}partition\PYGZus{}specificity} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{highlight\PYGZus{}genes} \PYG{n}{HIGHLIGHT\PYGZus{}GENES} \PYG{p}{[}\PYG{n}{HIGHLIGHT\PYGZus{}GENES} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{q\PYGZus{}thresh} \PYG{n}{Q\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fontsize} \PYG{n}{FONTSIZE}\PYG{p}{]}
                                 \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{custom\PYGZus{}palette} \PYG{n}{CUSTOM\PYGZus{}PALETTE} \PYG{p}{[}\PYG{n}{CUSTOM\PYGZus{}PALETTE} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}\PYG{p}{]}
                                 \PYG{n}{partition\PYGZus{}label} \PYG{n}{pvals\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat2}}\begin{optionlist}{3cm}
\item [partition\_label]  
\sphinxAtStartPar
Label for the partition being plotted, used in the plot title.
\item [pvals\_dir]  
\sphinxAtStartPar
Path to input CSV containing p\sphinxhyphen{}values and scores (Psi, Zeta, FDRs). CSV must have gene names as its index.
\item [save\_dir]  
\sphinxAtStartPar
Path where the output plot image will be saved.
\end{optionlist}


\subsubsection{Named Arguments}
\label{\detokenize{cli:named-arguments_repeat1}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}highlight\_genes]  
\sphinxAtStartPar
List of gene names to highlight and annotate on the plot (default: None).
\item [\sphinxhyphen{}\sphinxhyphen{}q\_thresh]  
\sphinxAtStartPar
Threshold for q\sphinxhyphen{}values (‘Psi q\sphinxhyphen{}value’ and ‘Zeta q\sphinxhyphen{}value’). Must be \textless{}= q\_thresh (default: 0.05).
\item [\sphinxhyphen{}\sphinxhyphen{}fontsize]  
\sphinxAtStartPar
Base font size for plot labels and text (default: 18).
\item [\sphinxhyphen{}\sphinxhyphen{}custom\_palette]  
\sphinxAtStartPar
List of 7 hex color codes to customize the color scheme. Order:
{[}‘significant by psi’, ‘significant by zeta’, ‘highlight genes’, ‘significant by both’, ‘circle markers’, ‘circle housekeeping genes’, ‘significant by neither’{]}. Default: None (uses built\sphinxhyphen{}in palette).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember plot\_partition\_specificity Genotype pvals\_entropy\_metrics\_Genotype\_WSBJ.csv output/ \textendash{}highlight\_genes Cwc22 \textendash{}fontsize 25

\end{description}
\end{quote}


\subsection{plot\_block\_specificity}
\label{\detokenize{cli:plot_block_specificity}}\begin{quote}

\sphinxAtStartPar
Generate a psi\_block vs. Psi scatter plot to visualize block\sphinxhyphen{}specific genes.

\sphinxAtStartPar
This function reads p\sphinxhyphen{}value data, colors genes based on their statistical
significance for Psi and psi\_block scores, and highlights the top genes
significant in both metrics. Allows for custom highlighting of a user\sphinxhyphen{}provided
gene list. Font size and color palette can be customized.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{plot\PYGZus{}block\PYGZus{}specificity} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{highlight\PYGZus{}genes} \PYG{n}{HIGHLIGHT\PYGZus{}GENES} \PYG{p}{[}\PYG{n}{HIGHLIGHT\PYGZus{}GENES} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{q\PYGZus{}thresh} \PYG{n}{Q\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fontsize} \PYG{n}{FONTSIZE}\PYG{p}{]}
                             \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{custom\PYGZus{}palette} \PYG{n}{CUSTOM\PYGZus{}PALETTE} \PYG{p}{[}\PYG{n}{CUSTOM\PYGZus{}PALETTE} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}\PYG{p}{]}
                             \PYG{n}{partition\PYGZus{}label} \PYG{n}{block\PYGZus{}label} \PYG{n}{pvals\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat3}}\begin{optionlist}{3cm}
\item [partition\_label]  
\sphinxAtStartPar
Label for the partition, used in the plot title.
\item [block\_label]  
\sphinxAtStartPar
Label for the block variable (e.g., a cell type or condition).
\item [pvals\_dir]  
\sphinxAtStartPar
Path to input CSV containing p\sphinxhyphen{}values and scores. CSV must have gene names as its index.
\item [save\_dir]  
\sphinxAtStartPar
Path where the output plot image will be saved.
\end{optionlist}


\subsubsection{Named Arguments}
\label{\detokenize{cli:named-arguments_repeat2}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}highlight\_genes]  
\sphinxAtStartPar
List of gene names to highlight and annotate on the plot (default: None).
\item [\sphinxhyphen{}\sphinxhyphen{}q\_thresh]  
\sphinxAtStartPar
Threshold for q\sphinxhyphen{}values (‘Psi q\sphinxhyphen{}value’ and ‘psi\_block q\sphinxhyphen{}value’). Must be \textless{}= q\_thresh (default: 0.05).
\item [\sphinxhyphen{}\sphinxhyphen{}fontsize]  
\sphinxAtStartPar
Base font size for plot labels and text (default: 18).
\item [\sphinxhyphen{}\sphinxhyphen{}custom\_palette]  
\sphinxAtStartPar
List of 7 hex color codes to customize the color scheme. Order:
{[}‘significant by psi’, ‘significant by psi\_block’, ‘highlight genes’, ‘significant by both’, ‘circle markers’, ‘circle housekeeping genes’, ‘significant by neither’{]}. Default: None (uses built\sphinxhyphen{}in palette).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember plot\_block\_specificity Genotype WSBJ pvals\_entropy\_metrics\_Genotype\_WSBJ.csv output/ \textendash{}highlight\_genes Cwc22 \textendash{}fontsize 25

\end{description}
\end{quote}


\subsection{plot\_sample\_counts}
\label{\detokenize{cli:plot_sample_counts}}\begin{quote}

\sphinxAtStartPar
Generate a bar plot showing the number of unique individuals per category and condition.

\sphinxAtStartPar
This function reads an AnnData object from an .h5ad file in backed mode,
calculates the number of unique individuals for each combination of a given
category and condition, and visualizes these counts as a grouped bar plot.
Font size can be customized.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{plot\PYGZus{}sample\PYGZus{}counts} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fontsize} \PYG{n}{FONTSIZE}\PYG{p}{]} \PYG{n}{h5ad\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir} \PYG{n}{sample\PYGZus{}id\PYGZus{}col} \PYG{n}{category\PYGZus{}col} \PYG{n}{condition\PYGZus{}col}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat4}}\begin{optionlist}{3cm}
\item [h5ad\_dir]  
\sphinxAtStartPar
Path to the input AnnData (.h5ad) file.
\item [save\_dir]  
\sphinxAtStartPar
Path to directory to save the output plot image.
\item [sample\_id\_col]  
\sphinxAtStartPar
Column name in \sphinxtitleref{.obs} that contains unique sample IDs.
\item [category\_col]  
\sphinxAtStartPar
Column name to use for the primary categories on the x\sphinxhyphen{}axis.
\item [condition\_col]  
\sphinxAtStartPar
Column name to use for grouping the bars (hue).
\end{optionlist}


\subsubsection{Named Arguments}
\label{\detokenize{cli:named-arguments_repeat3}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}fontsize]  
\sphinxAtStartPar
Base font size for plot labels and text (default: 18).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember plot\_sample\_counts test\_adata\_cwc22.h5ad \textasciitilde{}/ember\_test/output Mouse\_ID Genotype Sex \textendash{}fontsize 20

\end{description}
\end{quote}


\subsection{plot\_psi\_blocks}
\label{\detokenize{cli:plot_psi_blocks}}\begin{quote}

\sphinxAtStartPar
Generates and saves a bar plot of mean psi block values with error bars.

\sphinxAtStartPar
This function reads two CSV files from a specified directory: one for mean
psi block values and one for standard deviations. It plots the mean values
for a specific gene as a bar plot with corresponding standard deviation
error bars. Font size can be customized.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{plot\PYGZus{}psi\PYGZus{}blocks} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fontsize} \PYG{n}{FONTSIZE}\PYG{p}{]} \PYG{n}{gene\PYGZus{}name} \PYG{n}{partition\PYGZus{}label} \PYG{n}{psi\PYGZus{}block\PYGZus{}df\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat5}}\begin{optionlist}{3cm}
\item [gene\_name]  
\sphinxAtStartPar
Name of the gene (row) to select and plot from the CSV files.
\item [partition\_label]  
\sphinxAtStartPar
Partition label used to find the correct files (e.g., ‘Genotype’).
\item [psi\_block\_df\_dir]  
\sphinxAtStartPar
Directory containing the mean and std CSV files. Files must be named ‘mean\_Psi\_block\_df\_\{partition\_label\}.csv’ and ‘std\_Psi\_block\_df\_\{partition\_label\}.csv’.
\item [save\_dir]  
\sphinxAtStartPar
Path to directory to save the output plot image.
\end{optionlist}


\subsubsection{Named Arguments}
\label{\detokenize{cli:named-arguments_repeat4}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}fontsize]  
\sphinxAtStartPar
Base font size for plot labels and text (default: 18).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember plot\_psi\_blocks Cwc22 Genotype \textasciitilde{}/ember\_test/output/Psi\_block\_df/ \textasciitilde{}/ember\_test/output/figs \textendash{}fontsize 30

\end{description}
\end{quote}


\subsection{highly\_specific\_to\_partition}
\label{\detokenize{cli:highly_specific_to_partition}}\begin{quote}

\sphinxAtStartPar
Identifies significant and specific genes from an ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, Zeta, and q\sphinxhyphen{}values.
The resulting DataFrame is saved as “highly\_specific\_genes\_to\_\{partition\_label\}.csv”.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{highly\PYGZus{}specific\PYGZus{}to\PYGZus{}partition} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psi\PYGZus{}thresh} \PYG{n}{PSI\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{zeta\PYGZus{}thresh} \PYG{n}{ZETA\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{q\PYGZus{}thresh} \PYG{n}{Q\PYGZus{}THRESH}\PYG{p}{]}
                                   \PYG{n}{partition\PYGZus{}label} \PYG{n}{pvals\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat6}}\begin{optionlist}{3cm}
\item [partition\_label]  
\sphinxAtStartPar
Name of partition used to generate entropy metrics, used to label saved csv.
\item [pvals\_dir]  
\sphinxAtStartPar
Path to the input CSV file (must contain ‘Psi q\sphinxhyphen{}value’, ‘Zeta q\sphinxhyphen{}value’, ‘Psi’, and ‘Zeta’).
\item [save\_dir]  
\sphinxAtStartPar
Directory where the filtered results CSV will be saved.
\end{optionlist}


\subsubsection{Threshold Parameters}
\label{\detokenize{cli:threshold-parameters}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}psi\_thresh]  
\sphinxAtStartPar
Threshold for Psi values. Genes must have Psi \textgreater{} psi\_thresh (default: 0.5).
\item [\sphinxhyphen{}\sphinxhyphen{}zeta\_thresh]  
\sphinxAtStartPar
Threshold for Zeta values. Genes must have Zeta \textgreater{} zeta\_thresh (default: 0.5).
\item [\sphinxhyphen{}\sphinxhyphen{}q\_thresh]  
\sphinxAtStartPar
Threshold for q\sphinxhyphen{}values (‘Psi q\sphinxhyphen{}value’ and ‘Zeta q\sphinxhyphen{}value’). Must be \textless{}= q\_thresh (default: 0.05).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember highly\_specific\_to\_partition Genotype pvals\_entropy\_metrics\_Genotype.csv output/ \textendash{}psi\_thresh 0.6 \textendash{}zeta\_thresh 0.7

\end{description}
\end{quote}


\subsection{highly\_specific\_to\_block}
\label{\detokenize{cli:highly_specific_to_block}}\begin{quote}

\sphinxAtStartPar
Identifies significant and specific genes from an ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, psi\_block, and q\sphinxhyphen{}values.
REsultant genes are potential marker genes.
The resulting DataFrame is saved as “highly\_specific\_genes\_by\_\{partition\_label\}\_\{block\_label\}.csv”.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{highly\PYGZus{}specific\PYGZus{}to\PYGZus{}block} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psi\PYGZus{}thresh} \PYG{n}{PSI\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psi\PYGZus{}block\PYGZus{}thresh} \PYG{n}{PSI\PYGZus{}BLOCK\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{q\PYGZus{}thresh} \PYG{n}{Q\PYGZus{}THRESH}\PYG{p}{]}
                               \PYG{n}{partition\PYGZus{}label} \PYG{n}{block\PYGZus{}label} \PYG{n}{pvals\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat7}}\begin{optionlist}{3cm}
\item [partition\_label]  
\sphinxAtStartPar
Name of partition used to generate entropy metrics.
\item [block\_label]  
\sphinxAtStartPar
Name of block in partition used to generate entropy metrics.
\item [pvals\_dir]  
\sphinxAtStartPar
Path to the input CSV file (must contain ‘Psi q\sphinxhyphen{}value’, ‘psi\_block q\sphinxhyphen{}value’, ‘Psi’, and ‘psi\_block’).
\item [save\_dir]  
\sphinxAtStartPar
Directory where the filtered results CSV will be saved.
\end{optionlist}


\subsubsection{Threshold Parameters}
\label{\detokenize{cli:threshold-parameters_repeat1}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}psi\_thresh]  
\sphinxAtStartPar
Threshold for Psi values. Genes must have Psi \textgreater{} psi\_thresh (default: 0.5).
\item [\sphinxhyphen{}\sphinxhyphen{}psi\_block\_thresh]  
\sphinxAtStartPar
Threshold for psi\_block values. Genes must have psi\_block \textgreater{} psi\_block\_thresh (default: 0.5).
\item [\sphinxhyphen{}\sphinxhyphen{}q\_thresh]  
\sphinxAtStartPar
Threshold for q\sphinxhyphen{}values (‘Psi q\sphinxhyphen{}value’ and ‘psi\_block q\sphinxhyphen{}value’). Must be \textless{}= q\_thresh (default: 0.05).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember highly\_specific\_to\_block Genotype WSBJ pvals\_entropy\_metrics\_Genotype\_WSBJ.csv output/ \textendash{}psi\_thresh 0.6 \textendash{}psi\_block\_thresh 0.7

\end{description}
\end{quote}


\subsection{non\_specific\_to\_partition}
\label{\detokenize{cli:non_specific_to_partition}}\begin{quote}

\sphinxAtStartPar
Identifies significant but non\sphinxhyphen{}specific genes (potential housekeeping genes) from an ember generated
p\sphinxhyphen{}values/q\sphinxhyphen{}values CSV file based on thresholds for Psi, Zeta, and q\sphinxhyphen{}values.
Note: The Zeta filter is reversed, keeping Zeta \textless{} zeta\_thresh.
The resulting DataFrame is saved as “non\_specific\_genes\_to\_\{partition\_label\}.csv”.
\end{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ember} \PYG{n}{non\PYGZus{}specific\PYGZus{}to\PYGZus{}partition} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{psi\PYGZus{}thresh} \PYG{n}{PSI\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{zeta\PYGZus{}thresh} \PYG{n}{ZETA\PYGZus{}THRESH}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{q\PYGZus{}thresh} \PYG{n}{Q\PYGZus{}THRESH}\PYG{p}{]}
                                \PYG{n}{partition\PYGZus{}label} \PYG{n}{pvals\PYGZus{}dir} \PYG{n}{save\PYGZus{}dir}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{cli:positional-arguments_repeat8}}\begin{optionlist}{3cm}
\item [partition\_label]  
\sphinxAtStartPar
Name of partition used to generate entropy metrics, used to label saved csv.
\item [pvals\_dir]  
\sphinxAtStartPar
Path to the input CSV file (must contain ‘Psi q\sphinxhyphen{}value’, ‘Zeta q\sphinxhyphen{}value’, ‘Psi’, and ‘Zeta’).
\item [save\_dir]  
\sphinxAtStartPar
Directory where the filtered results CSV will be saved.
\end{optionlist}


\subsubsection{Threshold Parameters}
\label{\detokenize{cli:threshold-parameters_repeat2}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}psi\_thresh]  
\sphinxAtStartPar
Threshold for Psi values. Genes must have Psi \textgreater{} psi\_thresh (default: 0.5).
\item [\sphinxhyphen{}\sphinxhyphen{}zeta\_thresh]  
\sphinxAtStartPar
Threshold for Zeta values. Genes must have Zeta \textless{} zeta\_thresh (default: 0.5) to be considered non\sphinxhyphen{}specific.
\item [\sphinxhyphen{}\sphinxhyphen{}q\_thresh]  
\sphinxAtStartPar
Threshold for q\sphinxhyphen{}values (‘Psi q\sphinxhyphen{}value’ and ‘Zeta q\sphinxhyphen{}value’). Must be \textless{}= q\_thresh (default: 0.05).
\end{optionlist}
\begin{quote}
\begin{description}
\sphinxlineitem{Example:}
\sphinxAtStartPar
ember non\_specific\_to\_partition Genotype pvals\_entropy\_metrics\_Genotype.csv output/ \textendash{}psi\_thresh 0.6 \textendash{}zeta\_thresh 0.2

\end{description}
\end{quote}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\let\bigletter\sphinxstyleindexlettergroup
\bigletter{e}
\item\relax\sphinxstyleindexentry{ember\_py}\sphinxstyleindexpageref{ember_py:\detokenize{module-ember_py}}
\item\relax\sphinxstyleindexentry{ember\_py.generate\_pvals}\sphinxstyleindexpageref{ember_py:\detokenize{module-ember_py.generate_pvals}}
\item\relax\sphinxstyleindexentry{ember\_py.light\_ember}\sphinxstyleindexpageref{ember_py:\detokenize{module-ember_py.light_ember}}
\item\relax\sphinxstyleindexentry{ember\_py.plots}\sphinxstyleindexpageref{ember_py:\detokenize{module-ember_py.plots}}
\item\relax\sphinxstyleindexentry{ember\_py.top\_genes}\sphinxstyleindexpageref{ember_py:\detokenize{module-ember_py.top_genes}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}